File: yggdrasil_decision_forests/port/python/ydf/learner/custom_loss.cc
/*
 * Copyright 2022 Google LLC.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "ydf/learner/custom_loss.h"

#include <pybind11/numpy.h>

#include <cstddef>
#include <cstdint>
#include <memory>
#include <variant>

#include "absl/status/status.h"
#include "absl/status/statusor.h"
#include "absl/strings/string_view.h"
#include "absl/strings/substitute.h"
#include "absl/types/span.h"
#include "yggdrasil_decision_forests/learner/abstract_learner.h"
#include "yggdrasil_decision_forests/learner/gradient_boosted_trees/gradient_boosted_trees.h"
#include "yggdrasil_decision_forests/learner/gradient_boosted_trees/loss/loss_imp_custom_binary_classification.h"
#include "yggdrasil_decision_forests/learner/gradient_boosted_trees/loss/loss_imp_custom_multi_classification.h"
#include "yggdrasil_decision_forests/learner/gradient_boosted_trees/loss/loss_imp_custom_regression.h"
#include "ydf/utils/numpy_data.h"
#include "ydf/utils/pybind.h"
#include "yggdrasil_decision_forests/utils/logging.h"
#include "yggdrasil_decision_forests/utils/status_macros.h"

#ifdef _WIN32
typedef std::ptrdiff_t ssize_t;
#endif

namespace yggdrasil_decision_forests::port::python {

namespace {

// Pybind11 automatically casts Numpy arrays of shape (2, x) to pairs /
// sequences of length 2. This can lead to subtle user error, so we explicitly
// disallow such arrays.
absl::Status CheckGradientAndHessianShape(const py::sequence& seq) {
  if (py::isinstance<py::array>(seq)) {
    return absl::InvalidArgumentError(
        "The gradient_and_hessian function returned a numpy array, expected a "
        "Sequence of two numpy arrays.");
  }
  if (seq.size() != 2) {
    return absl::InvalidArgumentError(absl::Substitute(
        "The gradient_and_hessian function returned a sequence of length $0. "
        "Expected a sequence of length 2",
        seq.size()));
  }
  return absl::OkStatus();
}

absl::Status Check1DArrayShape(const py::array_t<float>& arr,
                               absl::string_view arr_name,
                               size_t expected_size) {
  py::buffer_info py_result_buf = arr.request();
  if (py_result_buf.ndim != 1 || py_result_buf.shape.size() != 1 ||
      py_result_buf.shape[0] != expected_size) {
    return absl::InvalidArgumentError(absl::Substitute(
        "The $0 must be a one-dimensional "
        "Numpy array of $1 elements. Got $2-dimensional array of $3 elements.",
        arr_name, expected_size, py_result_buf.ndim, py_result_buf.size));
  }
  return absl::OkStatus();
}

absl::Status Check2DArrayShape(const py::array_t<float>& arr,
                               absl::string_view arr_name,
                               size_t expected_size_x, size_t expected_size_y) {
  py::buffer_info py_result_buf = arr.request();
  if (py_result_buf.ndim != 2 || py_result_buf.shape.size() != 2 ||
      py_result_buf.shape[0] != expected_size_x ||
      py_result_buf.shape[1] != expected_size_y) {
    return absl::InvalidArgumentError(absl::Substitute(
        "The $0 must be a two-dimensional Numpy array of shape ($1, $2)."
        "Got $3-dimensional array of $4 elements.",
        arr_name, expected_size_x, expected_size_y, py_result_buf.ndim,
        py_result_buf.size));
  }
  return absl::OkStatus();
}

}  // namespace

absl::StatusOr<model::gradient_boosted_trees::CustomRegressionLossFunctions>
CCRegressionLoss::ToCustomRegressionLossFunctions() const {
  auto safe_initial_predictions = MakeSafeGilHolder(initial_predictions);
  auto cc_initial_predictions =
      [safe_initial_predictions](
          const absl::Span<const float>& labels,
          const absl::Span<const float>& weights) -> absl::StatusOr<float> {
    return SafePythonCall<absl::StatusOr<float>>(
        "initial_predictions", [&]() -> absl::StatusOr<float> {
          auto pylabels = SpanToSafeCopy(labels);
          auto pyweights = SpanToSafeCopy(weights);
          return (*safe_initial_predictions)(pylabels, pyweights);
        });
  };
  auto safe_loss = MakeSafeGilHolder(loss);
  auto cc_loss =
      [safe_loss](
          const absl::Span<const float>& labels,
          const absl::Span<const float>& predictions,
          const absl::Span<const float>& weights) -> absl::StatusOr<float> {
    return SafePythonCall<absl::StatusOr<float>>(
        "loss", [&]() -> absl::StatusOr<float> {
          auto pylabels = SpanToSafeCopy(labels);
          auto pyweights = SpanToSafeCopy(weights);
          auto pypredictions = SpanToSafeCopy(predictions);
          return (*safe_loss)(pylabels, pypredictions, pyweights);
        });
  };
  auto safe_gradient_and_hessian = MakeSafeGilHolder(gradient_and_hessian);
  auto cc_gradient_and_hessian =
      [safe_gradient_and_hessian](
          const absl::Span<const float>& labels,
          const absl::Span<const float>& predictions,
          absl::Span<float> gradient,
          absl::Span<float> hessian) mutable -> absl::Status {
    return SafePythonCall("gradient_and_hessian", [&]() -> absl::Status {
      auto pylabels = SpanToSafeCopy(labels);
      auto pypredictions = SpanToSafeCopy(predictions);
      auto py_result = (*safe_gradient_and_hessian)(pylabels, pypredictions);
      RETURN_IF_ERROR(CheckGradientAndHessianShape(py_result));
      auto py_gradient =
          py_result[0].cast<py::array_t<float, py::array::forcecast>>();
      auto py_hessian =
          py_result[1].cast<py::array_t<float, py::array::forcecast>>();
      RETURN_IF_ERROR(
          Check1DArrayShape(py_gradient, "gradient", gradient.size()));
      RETURN_IF_ERROR(Check1DArrayShape(py_hessian, "hessian", hessian.size()));

      StridedSpanFloat32 py_gradient_accessor(py_gradient);
      StridedSpanFloat32 py_hessian_accessor(py_hessian);
      for (ssize_t example_idx = 0; example_idx < gradient.size();
           ++example_idx) {
        // TODO: Consider removing this copy (also for other losses).
        gradient[example_idx] = -py_gradient_accessor[example_idx];
        hessian[example_idx] = -py_hessian_accessor[example_idx];
      }
      return absl::OkStatus();
    });
  };
  return model::gradient_boosted_trees::CustomRegressionLossFunctions{
      cc_initial_predictions, cc_loss, cc_gradient_and_hessian};
}

absl::StatusOr<
    model::gradient_boosted_trees::CustomBinaryClassificationLossFunctions>
CCBinaryClassificationLoss::ToCustomBinaryClassificationLossFunctions() const {
  // Wrap the Python functions in a std::shared_ptr with a custom deleter
  // that acquires the GIL before deleting the underlying Python function.
  // This avoids dangling references and ensures safe out-of-scope destruction.
  auto safe_initial_predictions = MakeSafeGilHolder(initial_predictions);
  auto cc_initial_predictions =
      [safe_initial_predictions](
          const absl::Span<const int32_t>& labels,
          const absl::Span<const float>& weights) -> absl::StatusOr<float> {
    return SafePythonCall<absl::StatusOr<float>>(
        "initial_predictions", [&]() -> absl::StatusOr<float> {
          auto pylabels = SpanToSafeCopy(labels);
          auto pyweights = SpanToSafeCopy(weights);
          return (*safe_initial_predictions)(pylabels, pyweights);
        });
  };
  auto safe_loss = MakeSafeGilHolder(loss);
  auto cc_loss =
      [safe_loss](
          const absl::Span<const int32_t>& labels,
          const absl::Span<const float>& predictions,
          const absl::Span<const float>& weights) -> absl::StatusOr<float> {
    return SafePythonCall<absl::StatusOr<float>>(
        "loss", [&]() -> absl::StatusOr<float> {
          auto pylabels = SpanToSafeCopy(labels);
          auto pyweights = SpanToSafeCopy(weights);
          auto pypredictions = SpanToSafeCopy(predictions);
          return (*safe_loss)(pylabels, pypredictions, pyweights);
        });
  };
  auto safe_gradient_and_hessian = MakeSafeGilHolder(gradient_and_hessian);
  auto cc_gradient_and_hessian =
      [safe_gradient_and_hessian](
          const absl::Span<const int32_t>& labels,
          const absl::Span<const float>& predictions,
          absl::Span<float> gradient,
          absl::Span<float> hessian) mutable -> absl::Status {
    return SafePythonCall("gradient_and_hessian", [&]() -> absl::Status {
      auto pylabels = SpanToSafeCopy(labels);
      auto pypredictions = SpanToSafeCopy(predictions);
      auto py_result = (*safe_gradient_and_hessian)(pylabels, pypredictions);
      RETURN_IF_ERROR(CheckGradientAndHessianShape(py_result));
      auto py_gradient =
          py_result[0].cast<py::array_t<float, py::array::forcecast>>();
      auto py_hessian =
          py_result[1].cast<py::array_t<float, py::array::forcecast>>();
      RETURN_IF_ERROR(
          Check1DArrayShape(py_gradient, "gradient", gradient.size()));
      RETURN_IF_ERROR(Check1DArrayShape(py_hessian, "hessian", hessian.size()));

      StridedSpanFloat32 py_gradient_accessor(py_gradient);
      StridedSpanFloat32 py_hessian_accessor(py_hessian);
      for (ssize_t example_idx = 0; example_idx < gradient.size();
           ++example_idx) {
        gradient[example_idx] = -py_gradient_accessor[example_idx];
        hessian[example_idx] = -py_hessian_accessor[example_idx];
      }
      return absl::OkStatus();
    });
  };
  return model::gradient_boosted_trees::CustomBinaryClassificationLossFunctions{
      cc_initial_predictions, cc_loss, cc_gradient_and_hessian};
}

absl::StatusOr<
    model::gradient_boosted_trees::CustomMultiClassificationLossFunctions>
CCMultiClassificationLoss::ToCustomMultiClassificationLossFunctions() const {
  auto safe_initial_predictions = MakeSafeGilHolder(initial_predictions);
  auto cc_initial_predictions =
      [safe_initial_predictions](
          const absl::Span<const int32_t>& labels,
          const absl::Span<const float>& weights,
          absl::Span<float> cc_initial_predictions) -> absl::Status {
    return SafePythonCall("initial_predictions", [&]() -> absl::Status {
      auto pylabels = SpanToSafeCopy(labels);
      auto pyweights = SpanToSafeCopy(weights);
      py::array_t<float> py_initial_predictions =
          (*safe_initial_predictions)(pylabels, pyweights);
      RETURN_IF_ERROR(Check1DArrayShape(py_initial_predictions,
                                        "initial_predictions",
                                        cc_initial_predictions.size()));
      StridedSpanFloat32 accessor(py_initial_predictions);
      for (ssize_t example_idx = 0; example_idx < cc_initial_predictions.size();
           ++example_idx) {
        cc_initial_predictions[example_idx] = accessor[example_idx];
      }
      return absl::OkStatus();
    });
  };
  auto safe_loss = MakeSafeGilHolder(loss);
  auto cc_loss =
      [safe_loss](
          const absl::Span<const int32_t>& labels,
          const absl::Span<const float>& predictions,
          const absl::Span<const float>& weights) -> absl::StatusOr<float> {
    int num_examples = labels.size();
    DCHECK_GT(num_examples, 0);
    int dimension = predictions.size() / num_examples;
    return SafePythonCall<absl::StatusOr<float>>(
        "loss", [&]() -> absl::StatusOr<float> {
          auto pylabels = SpanToSafeCopy(labels);
          auto pyweights = SpanToSafeCopy(weights);
          auto pypredictions = SpanToSafeCopy(predictions);
          pypredictions = pypredictions.reshape({num_examples, dimension});
          return (*safe_loss)(pylabels, pypredictions, pyweights);
        });
  };
  auto safe_gradient_and_hessian = MakeSafeGilHolder(gradient_and_hessian);
  auto cc_gradient_and_hessian =
      [safe_gradient_and_hessian](
          const absl::Span<const int32_t>& labels,
          const absl::Span<const float>& predictions,
          absl::Span<const absl::Span<float>> gradient,
          absl::Span<const absl::Span<float>> hessian) mutable -> absl::Status {
    int num_examples = labels.size();
    DCHECK_GT(num_examples, 0);
    int dimension = predictions.size() / num_examples;
    return SafePythonCall("gradient_and_hessian", [&]() -> absl::Status {
      auto pylabels = SpanToSafeCopy(labels);
      auto pypredictions = SpanToSafeCopy(predictions);
      pypredictions = pypredictions.reshape({num_examples, dimension});
      auto py_result = (*safe_gradient_and_hessian)(pylabels, pypredictions);
      RETURN_IF_ERROR(CheckGradientAndHessianShape(py_result));
      auto py_gradient =
          py_result[0].cast<py::array_t<float, py::array::forcecast>>();
      auto py_hessian =
          py_result[1].cast<py::array_t<float, py::array::forcecast>>();
      RETURN_IF_ERROR(
          Check2DArrayShape(py_gradient, "gradient", dimension, num_examples));
      RETURN_IF_ERROR(
          Check2DArrayShape(py_hessian, "hessian", dimension, num_examples));

      auto py_gradient_unchecked = py_gradient.unchecked<2>();
      auto py_hessian_unchecked = py_hessian.unchecked<2>();
      for (ssize_t example_idx = 0; example_idx < num_examples; ++example_idx) {
        for (int grad_idx = 0; grad_idx < dimension; grad_idx++) {
          gradient[grad_idx][example_idx] =
              -py_gradient_unchecked(grad_idx, example_idx);
          hessian[grad_idx][example_idx] =
              -py_hessian_unchecked(grad_idx, example_idx);
        }
      }
      return absl::OkStatus();
    });
  };
  return model::gradient_boosted_trees::CustomMultiClassificationLossFunctions{
      cc_initial_predictions, cc_loss, cc_gradient_and_hessian};
}

absl::Status ApplyCustomLoss(const CCCustomLoss& custom_loss,
                             model::AbstractLearner* learner) {
  if (std::holds_alternative<std::monostate>(custom_loss)) {
    return absl::OkStatus();
  }
  auto* gbt_learner =
      dynamic_cast<model::gradient_boosted_trees::GradientBoostedTreesLearner*>(
          learner);
  if (!gbt_learner) {
    return absl::InvalidArgumentError(
        "Custom losses are only compatible with Gradient Boosted Trees.");
  }
  if (std::holds_alternative<CCRegressionLoss>(custom_loss)) {
    ASSIGN_OR_RETURN(auto loss_func, std::get<CCRegressionLoss>(custom_loss)
                                         .ToCustomRegressionLossFunctions());
    gbt_learner->SetCustomLossFunctions(loss_func);
  } else if (std::holds_alternative<CCBinaryClassificationLoss>(custom_loss)) {
    ASSIGN_OR_RETURN(auto loss_func,
                     std::get<CCBinaryClassificationLoss>(custom_loss)
                         .ToCustomBinaryClassificationLossFunctions());
    gbt_learner->SetCustomLossFunctions(loss_func);
  } else if (std::holds_alternative<CCMultiClassificationLoss>(custom_loss)) {
    ASSIGN_OR_RETURN(auto loss_func,
                     std::get<CCMultiClassificationLoss>(custom_loss)
                         .ToCustomMultiClassificationLossFunctions());
    gbt_learner->SetCustomLossFunctions(loss_func);
  } else {
    NOTREACHED();
  }
  return absl::OkStatus();
}

}  // namespace yggdrasil_decision_forests::port::python

File: yggdrasil_decision_forests/port/python/ydf/learner/custom_loss.py
# Copyright 2022 Google LLC.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""Implementations for custom loss containers."""

import dataclasses
import enum
from typing import Any, Callable, Tuple

import numpy as np
import numpy.typing as npt

from yggdrasil_decision_forests.model import abstract_model_pb2
from ydf.cc import ydf


class Activation(enum.Enum):
  """Activation functions for custom losses.

  Not all activation functions are supported for all custom losses. Activation
  function IDENTITY (i.e., no activation function applied) is always supported.
  """

  IDENTITY = "IDENTITY"
  SIGMOID = "SIGMOID"
  SOFTMAX = "SOFTMAX"


@dataclasses.dataclass(frozen=True)
class AbstractCustomLoss:
  """Abstract Base class for custom losses."""

  activation: Any

  initial_predictions: Callable[
      [npt.NDArray[Any], npt.NDArray[np.float32]],
      Any,
  ]
  loss: Callable[
      [
          npt.NDArray[Any],
          npt.NDArray[np.float32],
          npt.NDArray[np.float32],
      ],
      np.float32,
  ]
  gradient_and_hessian: Callable[
      [npt.NDArray[Any], npt.NDArray[np.float32]],
      Tuple[npt.NDArray[np.float32], npt.NDArray[np.float32]],
  ]

  may_trigger_gc: bool = True

  def check_is_compatible_task(self, task: abstract_model_pb2.Task) -> None:
    raise NotImplementedError("Not implemented")

  def _to_cc(self):
    raise NotImplementedError("Not implemented")


@dataclasses.dataclass(frozen=True)
class RegressionLoss(AbstractCustomLoss):
  """A user-provided loss function for regression problems.

  Loss functions may never reference their arguments outside after returning:
  Bad:
  ```
  mylabels = None
  def initial_predictions(labels, weights):
    nonlocal mylabels
    mylabels = labels  # labels is now referenced outside the function
  ```
  Good:
  ```
  mylabels = None
  def initial_predictions(labels, weights):
    nonlocal mylabels
    mylabels = np.copy(labels)  # mylabels is a copy, not a reference.
  ```

  Attributes:
    initial_predictions: The bias / initial predictions of the GBT model.
      Receives the label values and the weights, outputs the initial prediction
      as a float.
    loss: The loss function controls the early stopping. The loss function
      receives the labels, the current predictions and the current weights and
      must output the loss as a float. Note that the predictions provided to the
      loss functions have not yet had an activation function applied to them.
    gradient_and_hessian: Gradient and hessian of the current predictions. Note
      that only the diagonal of the hessian must be provided. Receives as input
      the labels and the current predictions (without activation) and returns a
      tuple of the gradient and the hessian.
    activation: Activation function to be applied to the model. Regression
      models are expected to return a value in the same space as the labels
      after applying the activation function.
    may_trigger_gc: If True (default), YDF may trigger Python's garbage
      collection to determine if a Numpy array that is backed by YDF-internal
      data is used after its lifetime has ended. If False, checks for illegal
      memory accesses are disabled. This can be useful when training many small
      models or if the observed impact of triggering GC is large. If
      `may_trigger_gc=False`, it is very important that the user validate
      manuallythat no memory leakage occurs.
  """

  initial_predictions: Callable[
      [npt.NDArray[np.float32], npt.NDArray[np.float32]],
      np.float32,
  ]
  loss: Callable[
      [
          npt.NDArray[np.float32],
          npt.NDArray[np.float32],
          npt.NDArray[np.float32],
      ],
      np.float32,
  ]
  gradient_and_hessian: Callable[
      [npt.NDArray[np.float32], npt.NDArray[np.float32]],
      Tuple[npt.NDArray[np.float32], npt.NDArray[np.float32]],
  ]

  activation: Activation

  def __post_init__(self):
    if self.activation != Activation.IDENTITY:
      raise ValueError(
          "Only activation function IDENTITY is supported for RegressionLoss."
      )

  def check_is_compatible_task(self, task: abstract_model_pb2.Task) -> None:
    """Raises an error if the given task is incompatible with this loss type."""
    if task != abstract_model_pb2.REGRESSION:
      raise ValueError(
          "A RegressionLoss is only compatible with REGRESSION"
          f" tasks. Received task {abstract_model_pb2.Task.Name(task)}"
      )

  def _to_cc(self):
    return ydf.CCRegressionLoss(
        self.initial_predictions,
        self.loss,
        self.gradient_and_hessian,
        self.may_trigger_gc,
    )


@dataclasses.dataclass(frozen=True)
class BinaryClassificationLoss(AbstractCustomLoss):
  """A user-provided loss function for binary classification problems.

  Note that the labels are binary but 1-based, i.e. the positive class is 2, the
  negative class is 1.

  Loss functions may never reference their arguments outside after returning:
  Bad:
  ```
  mylabels = None
  def initial_predictions(labels, weights):
    nonlocal mylabels
    mylabels = labels  # labels is now referenced outside the function
  ```
  Good:
  ```
  mylabels = None
  def initial_predictions(labels, weights):
    nonlocal mylabels
    mylabels = np.copy(labels)  # mylabels is a copy, not a reference.
  ```

  Attributes:
    initial_predictions: The bias / initial predictions of the GBT model.
      Receives the label values and the weights, outputs the initial prediction
      as a float.
    loss: The loss function controls the early stopping. The loss function
      receives the labels, the current predictions and the current weights and
      must output the loss as a float. Note that the predictions provided to the
      loss functions have not yet had an activation function applied to them.
    gradient_and_hessian: Gradient and hessian of the current predictions. Note
      that only the diagonal of the hessian must be provided. Receives as input
      the labels and the current predictions (without activation). Returns a
      tuple of the gradient and the hessian.
    activation: Activation function to be applied to the model. Binary
      classification models are expected to return a probability after applying
      the activation function.
    may_trigger_gc: If True (default), YDF may trigger Python's garbage
      collection to determine if an Numpy array that is backed by YDF-internal
      data is used after its lifetime has ended. If False, checks for illegal
      memory accesses are disabled. Setting this parameter to False is
      dangerous, since illegal memory accesses will no longer be detected.
  """

  initial_predictions: Callable[
      [npt.NDArray[np.int32], npt.NDArray[np.float32]],
      np.float32,
  ]
  loss: Callable[
      [
          npt.NDArray[np.int32],
          npt.NDArray[np.float32],
          npt.NDArray[np.float32],
      ],
      np.float32,
  ]
  gradient_and_hessian: Callable[
      [npt.NDArray[np.int32], npt.NDArray[np.float32]],
      Tuple[npt.NDArray[np.float32], npt.NDArray[np.float32]],
  ]

  activation: Activation

  def __post_init__(self):
    if self.activation not in [Activation.IDENTITY, Activation.SIGMOID]:
      raise ValueError(
          "Only activation functions IDENTITY and SIGMOID are supported for"
          " BinaryClassificationLoss."
      )

  def check_is_compatible_task(self, task: abstract_model_pb2.Task) -> None:
    """Raises an error if the given task is incompatible with this loss type."""
    if task != abstract_model_pb2.CLASSIFICATION:
      raise ValueError(
          "A BinaryClassificationLoss is only compatible with CLASSIFICATION"
          f" tasks. Received task {abstract_model_pb2.Task.Name(task)}"
      )

  def _to_cc(self):
    return ydf.CCBinaryClassificationLoss(
        self.initial_predictions,
        self.loss,
        self.gradient_and_hessian,
        self.may_trigger_gc,
    )


@dataclasses.dataclass(frozen=True)
class MultiClassificationLoss(AbstractCustomLoss):
  """A user-provided loss function for multi-class problems.

  Note that the labels are but 1-based. Predictions are given in an 2D
  array with one row per example. Initial predictions, gradient and
  hessian are expected for each class, e.g. for a 3-class classification
  problem, output 3 gradients and hessians per class.

  Loss functions may never reference their arguments outside after returning:
  Bad:
  ```
  mylabels = None
  def initial_predictions(labels, weights):
    nonlocal mylabels
    mylabels = labels  # labels is now referenced outside the function
  ```
  Good:
  ```
  mylabels = None
  def initial_predictions(labels, weights):
    nonlocal mylabels
    mylabels = np.copy(labels)  # mylabels is a copy, not a reference.
  ```

  Attributes:
    initial_predictions: The bias / initial predictions of the GBT model.
      Receives the label values and the weights, outputs the initial prediction
      as an array of floats (one initial prediction per class).
    loss: The loss function controls the early stopping. The loss function
      receives the labels, the current predictions and the current weights and
      must output the loss as a float. Note that the predictions provided to the
      loss functions have not yet had an activation function applied to them.
    gradient_and_hessian: Gradient and hessian of the current predictions with
      respect to each class. Note that only the diagonal of the hessian must be
      provided. Receives as input the labels and the current predictions
      (without activation). Returns a tuple of the gradient and the hessian.
      Both gradient and hessian must be arrays of shape (num_classes,
      num_examples).
    activation: Activation function to be applied to the model. Multi-class
      classification models are expected to return a probability distribution
      over the classes after applying the activation function.
    may_trigger_gc: If True (default), YDF may trigger Python's garbage
      collection to determine if an Numpy array that is backed by YDF-internal
      data is used after its lifetime has ended. If False, checks for illegal
      memory accesses are disabled. Setting this parameter to False is
      dangerous, since illegal memory accesses will no longer be detected.
  """

  initial_predictions: Callable[
      [npt.NDArray[np.int32], npt.NDArray[np.float32]],
      npt.NDArray[np.float32],
  ]
  loss: Callable[
      [
          npt.NDArray[np.int32],
          npt.NDArray[np.float32],
          npt.NDArray[np.float32],
      ],
      np.float32,
  ]
  gradient_and_hessian: Callable[
      [npt.NDArray[np.int32], npt.NDArray[np.float32]],
      Tuple[npt.NDArray[np.float32], npt.NDArray[np.float32]],
  ]

  activation: Activation

  def __post_init__(self):
    if (
        self.activation != Activation.IDENTITY
        and self.activation != Activation.SOFTMAX
    ):
      raise ValueError(
          "Only activation functions IDENTITY and SOFTMAX are supported for"
          " MultiClassificationLoss."
      )

  def check_is_compatible_task(self, task: abstract_model_pb2.Task) -> None:
    """Raises an error if the given task is incompatible with this loss type."""
    if task != abstract_model_pb2.CLASSIFICATION:
      raise ValueError(
          "A MultiClassificationLoss is only compatible with CLASSIFICATION"
          f" tasks. Received task {abstract_model_pb2.Task.Name(task)}"
      )

  def _to_cc(self):
    return ydf.CCMultiClassificationLoss(
        self.initial_predictions,
        self.loss,
        self.gradient_and_hessian,
        self.may_trigger_gc,
    )
